fenicsprecice/adapter_core.py

```diff
@@ -92,96 +92,183 @@ def filter_point_sources(point_sources, filter_out):
+            for lid in lids:
+                result_vals.append([vertex_vals[lid], vertex_vals[lid+1]])
+        else:  # function space is FunctionSpace (scalar)
+            for lid in lids:
+                result_vals.append(vertex_vals[lid])
     else:
-        raise Exception("Cannot handle data type %s" % type(data))
+        result_vals = np.array([])
+
+    return np.array(result_vals)
 
 
-def get_coupling_boundary_vertices(mesh_fenics, coupling_subdomain, fenics_dimensions, dimensions):
+def get_fenics_coupling_boundary_vertices(function_space, coupling_subdomain):
     """
-    Extracts vertices from a given mesh which lie on the  given coupling domain.
+    Extracts vertices which this rank owns from a given function space which lie on the given coupling domain.
 
     Parameters
     ----------
-    mesh_fenics : FEniCS Mesh
-        Mesh of complete domain.
-    coupling_subdomain : FeniCS Domain
+    function_space : FEniCS function space
+        Function space on which the finite element problem definition lives.
+    coupling_subdomain : FEniCS Domain
         Subdomain consists of only the coupling interface region.
-    fenics_dimensions : int
-        Dimensions of FEniCS case setup.
-    dimensions : int
-        Dimensions of coupling case setup.
 
     Returns
     -------
-    fenics_vertices : numpy array
-        Array consisting of all vertices lying on the coupling interface.
-    coordinates : array_like
-        The coordinates of the vertices in a numpy array [N x D] where
+    fenics_gids : numpy array
+        Array of global indices of vertices on the coupling interface as seen by this rank.
+    fenics_coords : array_like
+        The coordinates of the vertices on the coupling interface as seen by this rank, in a numpy array [N x D] where
         N = number of vertices and D = dimensions of geometry.
-    n : int
-        Number of vertices on the coupling interface.
     """
-    n = 0
-    fenics_vertices = []
-    vertices_x = []
-    vertices_y = []
-    vertices_z = []
 
     if not issubclass(type(coupling_subdomain), SubDomain):
         raise Exception("No correct coupling interface defined! Given coupling domain is not of type dolfin Subdomain")
 
-    for v in dolfin.vertices(mesh_fenics):
+    # Get mesh from FEniCS function space
+    mesh = function_space.mesh()
+
+    # Get coordinates and global IDs of all vertices of the mesh  which lie on the coupling boundary.
+    # These vertices include shared (owned + unowned) and non-shared vertices in a parallel setting
+    fenics_gids, fenics_coords = [], []
+    for v in vertices(mesh):
+        if coupling_subdomain.inside(v.point(), True):
+            fenics_gids.append(v.global_index())
+            fenics_coords.append([v.x(0), v.x(1)])
+
+    return np.array(fenics_gids), np.array(fenics_coords)
+
+
+def get_owned_coupling_boundary_vertices(function_space, coupling_subdomain):
```

Comment:
From looking at the function name and documentation I do not really see a big difference between this function and `get_fenics_coupling_boundary_vertices`. Is it possible to provide one function serving both purposes? My impression:

* `get_fenics_coupling_boundary_vertices` returns `fenics_gids` and `fenics_coords` on `coupling_subdomain` owned by rank.
* `get_owned_coupling_boundary_vertices` returns `fenics_gids`, **`fenics_lids`**, and `fenics_coords`  on `coupling_subdomain` owned by rank.

This is almost the same. Can we not just use `get_owned_coupling_boundary_vertices` everywhere? Is it too costly to get `fenics_lids`, if they are not needed?